pkg/image/api/types.go

```diff
@@ -60,6 +77,41 @@ type ImageLayer struct {
 	Size int64
 }
 
+// SignatureValidationResult holds a result of signature validation.
+type SignatureValidationResult string
+
+// ImageSignature holds a signature of an image.
+type ImageSignature struct {
+	// Describes a type of stored blob. Provided by caller.
+	Type string
+	// An opaque binary string which is an image's signature.
+	Content []byte
+	// A time of last server-side validation.
+	ValidatedOn unversioned.Time
+	// Result of validation performed at ValidatedOn.
+	ValidationResult SignatureValidationResult
+
+	// At most one of the following pointers can point to details. If the signature type isn't recognized,
+	// they will all be nil.
+
+	// Details for signature of type "atomic".
+	AtomicSignatureDetails *AtomicSignatureDetails
+	// For any future supported signature type, add additional pointer to its own details struct.
+}
+
+// AtomicSignatureDetails holds metadata of recognized signature of type "atomic". They are parsed from
+// signature's content.
+type AtomicSignatureDetails struct {
+	// Digest of signed manifest.
+	SignedImageName string
+	// Docker image reference
+	SignedDockerImageReference string
+	// Optional, if specified, it is an identity of signer.
+	Creator string
```

Comment:
We would call this the Provider or the UserAgent in general.  It's fine if
it is removed.

On Thu, May 19, 2016 at 11:59 AM, Miloslav Trmač <notifications@github.com>
wrote:

> In pkg/image/api/types.go
> <https://github.com/openshift/origin/pull/8371#discussion_r63905790>:
>
> > +	// they will all be nil.
> > +
> > +	// Details for signature of type "atomic".
> > +	AtomicSignatureDetails *AtomicSignatureDetails
> > +	// For any future supported signature type, add additional pointer to its own details struct.
> > +}
> > +
> > +// AtomicSignatureDetails holds metadata of recognized signature of type "atomic". They are parsed from
> > +// signature's content.
> > +type AtomicSignatureDetails struct {
> > +	// Digest of signed manifest.
> > +	SignedImageName string
> > +	// Docker image reference
> > +	SignedDockerImageReference string
> > +	// Optional, if specified, it is an identity of signer.
> > +	Creator string
>
> It doesn't necessarily refer to person, it may be a name/version of
> signing service.
>
> It *never* refers to a person, it is name/version of a signing
> *implementation* to allow for future workarounds of past bugs. See projectatomic/skopeo#28
> (comment)
> <https://github.com/projectatomic/skopeo/pull/28#discussion_r57299345>
> and replies for the details.
>
> “Who signed this” is not embedded in the signature; it is implicit in the
> key ID within the signature blob, and “which key should have signed this”
> can be derived from SignedDockerImageReference as discussed above.
>
> IMHO it is entirely useless to display this to the user, and therefore to
> also expose it in the OpenShift API (sorry, Michal, my fault for not
> noticing earlier.)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8371/files/a49f7a114201f3d8868dba2c837f84b85045e79c#r63905790>
>